L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,339 @@
+
+These two cases indicate the following design requirements:
+1. Support both use cases above. We cannot assume there are bootstrapping key materials. 
+2. The credentials reloading C core API needs to be simple, so that it can be easily wrapped by different languages. The credentials reloading C++ API needs to be user-friendly and simple to use. If possible, the API shall avoid complex asynchronous programming (schedule, cancel, mutex, etc.) by the gRPC users.
+3. The credentials reloading should be efficient. We shall reload only when needed rather than reloading per handshake.
+
+Based on the requirements, we are going to propose a new distributor/provider API that will meet these requirements. 
+The distributor is responsible for caching the credentials and distributing them in each handshake. 
+The users will need to specify a particular provider implementation when creating their credentials, and we will provide one implementation for reloading from files.
+If that doesn't meet users' requirements, they could implement their own providers.
+
+Besides major credential reloading change, we are also planning to:
+1. make the server authorization original designed for client side only applicable to both the client side and the server side. 
+This is based on some recent user requirement, and it may also be used as the CRL check since CRL is not supported in gRPC core. 
+2. add the options to configure the min/max TLS version that will be used in handshake
+
+### Related Proposals:
+
+[L46: Create a new TLS credential API that supports SPIFFE mutual TLS](https://github.com/grpc/proposal/pull/98)
+
+## Proposal
+
+### Changes around TLS credentials and Options
+This part of the proposal introduces changes made to existing TLS credential APIs. Changes include:
+1. Replacing all the occurrence of "*_server_authorization_*" to "*_authorization_"
+2. Removing `grpc_tls_key_materials_config`, `grpc_tls_credential_reload_config` and `grpc_tls_credential_reload_arg` 
+3. Adding API `grpc_tls_credentials_options_set_certificate_provider`, `grpc_tls_server_credentials_set_min_tls_version` and `grpc_tls_server_credentials_set_max_tls_version`
+4. Adding API `grpc_tls_credentials_options_set_root_cert_name` and `grpc_tls_credentials_options_set_identity_cert_name`
+
+Here is how the API would look like after these changes. For simplicity, all the comments are omitted.
+
+```c
+/** --- TLS channel/server credentials ---
+ * It is used for experimental purpose for now and subject to change. */
+
+typedef struct grpc_tls_error_details grpc_tls_error_details;
+
+typedef struct grpc_tls_authorization_check_config grpc_tls_authorization_check_config;
+
+typedef struct grpc_tls_credentials_options grpc_tls_credentials_options;
+
+GRPCAPI grpc_tls_credentials_options* grpc_tls_credentials_options_create(void);
+
+GRPCAPI int grpc_tls_credentials_options_set_cert_request_type(
```

Comment:
please specify the default value if it is not called. Same for `grpc_tls_credentials_options_set_server_verification_option` below